doc/adr/0004-nats-headers.md

```diff
@@ -8,38 +8,73 @@ Accepted
+parsing.
 
 The salient points of the HTTP header specification are:
 
-- Each header field consists of a case-insensitive field name followed by a colon (`:`), optional leading whitespace, the field value, and optional trailing whitespace.
+- Each header field consists of a case-insensitive field name followed by a
+  colon (`:`), optional leading whitespace, the field value, and optional
+  trailing whitespace.
 - No spaces are allowed between the header field name and colon.
 - Field value may be preceded or followed by optional whitespace.
-- The specification may allow any number of strange things like comments/tokens etc.
+- The specification may allow any number of strange things like comments/tokens
+  etc.
 - The keys can repeat.
 
-More specifically from [rfc822](https://www.ietf.org/rfc/rfc822.txt) Section 3.1.2:
-> Once a field has been unfolded, it may be viewed as being composed of a field-name followed by a colon (":"), followed by a field-body, and  terminated  by  a  carriage-return/line-feed.
-> The  field-name must be composed of printable ASCII characters (i.e., characters that  have  values  between  33.  and  126., decimal, except colon).  The field-body may be composed of any ASCII characters, except CR or LF.  (While CR and/or LF may be present  in the actual text, they are removed by the action of unfolding the field.)
+More specifically from [rfc822](https://www.ietf.org/rfc/rfc822.txt) Section
+3.1.2:
 
-The reference NATS Go client uses the `http.Header` which internally uses `textproto.MIMEHeader` infrastructure for serializing NATS headers. Other language clients may not be able to reuse header serialization infrastructure outside of the context of an HTTP request. Furthermore, the implementation of the native header library may diverge from the Go `http.Header` while still being standard.
+> Once a field has been unfolded, it may be viewed as being composed of a
+> field-name followed by a colon (":"), followed by a field-body, and terminated
+> by a carriage-return/line-feed. The field-name must be composed of printable
+> ASCII characters (i.e., characters that have values between 33. and 126.,
+> decimal, except colon). The field-body may be composed of any ASCII
+> characters, except CR or LF. (While CR and/or LF may be present in the actual
+> text, they are removed by the action of unfolding the field.)
+
+The reference NATS Go client uses the `http.Header` which internally uses
+`textproto.MIMEHeader` infrastructure for serializing NATS headers. Other
+language clients may not be able to reuse header serialization infrastructure
+outside of the context of an HTTP request. Furthermore, the implementation of
+the native header library may diverge from the Go `http.Header` while still
+being standard.
+
+For example, the Go `http.Header` library, performs the following formatting
+which while not part of the spec, as header field names are case-insensitive,
+become a requirement in NATS header fields:
```

Comment:
This is not correct. We no longer use MIME canonicalization in the Go client. Here is an example of setting a header in the Go client:
```
m := nats.NewMsg(subj)
m.Header.Set("ivan", "kozlovic")
nc.PublishMsg(m)
```
and how it is received in the server:
```
<<- MSG_PAYLOAD: ["NATS/1.0\r\nivan: kozlovic\r\n\r\n"]
```
As you can see, the key has not been canonicalized.